### PR TITLE
Add doc for Smartmeter Service

### PR DESCRIPTION
**Pull request by xBlade58**

### CODE REVIEW

File: documentation/implementation/adapter-and-concentrator/Adapter_Concentrator.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+Here is a code snippet where the message is formatted and sent to the MQTT broker in the Cloud.
+
+```python
+def subscribe(client_sub: mqtt_client.Client, client_pub: mqtt_client.Client):
+    def on_message(client_sub, userdata, msg):
+        msg = json.loads(msg.payload.decode())
+        final_msg = adapt(msg) # formats message
+        json_msg = json.dumps(final_msg)
+        print('Sedning final msg: \n', json_msg)
```
Comment (bellerstefan) [inline documentation/implementation/adapter-and-concentrator/Adapter_Concentrator.md:103]:
```suggestion
        print('Sending final msg: \n', json_msg)
```

File: documentation/implementation/adapter-and-concentrator/Adapter_Concentrator.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        result = client_pub.publish(topic, json_msg)
+        status = result[0]
+        if status != 0:
+            print(f'Failed to send message to cloud')
+
+```
+
+In regard to choosing the technology for the MQTT broker, we decided to go with [EMQX](https://www.emqx.io/). In our evaluation, we considered brokers like HiveMQ, Mosquitto, RabbitMQ. EMQX, however, stood out thanks to its high performance and scalability benchmarks.
+
+TODO: add links to sources
```
Comment (bellerstefan) [inline documentation/implementation/adapter-and-concentrator/Adapter_Concentrator.md:113]: Fix TODO

File: documentation/implementation/smartmeter-service/Smartmeter_Service.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+![domain_model](../images/Domain_Model.png)
+
+For modelling the domain, we adhered to the data model as proposed by the specification. One crucial relation is depicted by the `MeterReading` containing multiple `PropertyValues`. Every PropertyValue is defined by a _OperationalPropertyDef_. An example for an identifier of such a property definition would be `52.7.0`, which holds the **Instantaneous voltage in phase L2**.
+
+## Hexagonal Architecture
+
+Our goal was to build the Smartmeter service in a way that enables different external components to interact with our business logic, while keeping the business logic isolated from any external dependencies. So, the decision fell to a `Hexagonal Architecture`, also commonly known as `Ports and Adapters` pattern.
+
+Consequently, we divided the implementation of the Smartmeter service into two parts. The first part is the _application (inner hexagon)_ that handles the business logic through the domain model and ports. The second part is represented by the _adapters (outer hexagon)_ that use the ports to interact with the business logic. In the illustration below, an extract ofr our hexagonal architecture can be seen.
```
Comment (bellerstefan) [inline documentation/implementation/smartmeter-service/Smartmeter_Service.md:22]:
```suggestion
Consequently, we divided the implementation of the Smartmeter service into two parts. The first part is the _application (inner hexagon)_ that handles the business logic through the domain model and ports. The second part is represented by the _adapters (outer hexagon)_ that use the ports to interact with the business logic. In the illustration below, an extract of our hexagonal architecture can be seen.
```

File: documentation/implementation/smartmeter-service/Smartmeter_Service.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+An adapter allows interaction through a particular port using a particular technology.
+
+#### MQTT Adapter
+
+One important adapter is the `MQTT-Adapter`, which is basically a subscriber to the EMQX-Broker. Whenever it receives a new meter reading message, the message is parsed and the `CreateMeterReadingUseCase` is used to store the data.
+
+#### TimescaleDB Adapter
+
+The `TimescaleDB` adapters implement the repository interface to save and fetch from our Timescale database. Internally, they delegete most of the requests to the respective JPA repositories, which are provided by Spring to simplify basic operations to the underlying database.
```
Comment (bellerstefan) [inline documentation/implementation/smartmeter-service/Smartmeter_Service.md:46]:
```suggestion
The `TimescaleDB` adapters implement the repository interface to save and fetch from our Timescale database. Internally, they delegate most of the requests to the respective JPA repositories, which are provided by Spring to simplify basic operations to the underlying database.
```

File: documentation/implementation/smartmeter-service/Smartmeter_Service.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#### Redis Adapter
+
+Apart from that we also implemented a `Redis-Adapter` that listens for `HouseholdEvents`. For example, after a `MeterAssignedEvent` is sent by the Household microservice, the Smartmeter service updates the respective `householdId` field of the Smart Meter.
+
+Finally, we want to note that our objective was not to develop a fully Domain-Driven-Design pattern, despite the fact that you might recognise some DDD patterns in the implementation. We rather wanted to develop a "clean" Hexagonal Architecture.
+
+## Technology Stack
+
+- Spring Boot
+- HIbernate
```
Comment (bellerstefan) [inline documentation/implementation/smartmeter-service/Smartmeter_Service.md:57]: intended?

**Review by bellerstefan — COMMENTED** [review]
How would we like to write smart meter?

File: documentation/implementation/smartmeter-service/Smartmeter_Service.md (hunk 5 above)
Comment (xBlade58) [reply documentation/implementation/smartmeter-service/Smartmeter_Service.md:57]: No thanks haha

**Review by bellerstefan — APPROVED** [review]
all good

**Pull request merged**